thavalon-server/src/game/engine.rs

```diff
@@ -0,0 +1,761 @@
+    fn upvotes<'a>(&'a self) -> impl Iterator<Item = PlayerId> + 'a {
+        self.votes
+            .iter()
+            .filter_map(|(player, vote)| if *vote { Some(*player) } else { None })
+    }
+
+    /// Players who downvoted the proposal
+    fn downvotes<'a>(&'a self) -> impl Iterator<Item = PlayerId> + 'a {
+        self.votes
+            .iter()
+            .filter_map(|(player, vote)| if !*vote { Some(*player) } else { None })
+    }
+}
+
+impl<'a> fmt::Display for Proposal<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.players
+            .iter()
+            .format_with(", ", |id, f| f(&self.game.name(*id)))
+            .fmt(f)?;
+        write!(f, " (proposed by {})", self.game.name(self.proposer))
+    }
+}
+
+#[test]
+fn test_is_failure() {
+    use super::GameSpec;
+    let spec = GameSpec::for_players(5);
+
+    assert!(!is_failure(
+        spec,
+        1,
+        &[Card::Success, Card::Success, Card::Success]
+    ));
+    assert!(!is_failure(
+        spec,
+        1,
+        &[Card::Success, Card::Fail, Card::Reverse]
+    ));
+    assert!(!is_failure(
+        spec,
+        1,
+        &[Card::Success, Card::Reverse, Card::Reverse]
+    ));
+
+    assert!(is_failure(
+        spec,
+        1,
+        &[Card::Success, Card::Success, Card::Fail]
+    ));
+    assert!(is_failure(
+        spec,
+        1,
+        &[Card::Success, Card::Fail, Card::Fail]
+    ));
+    assert!(is_failure(spec, 1, &[Card::Fail, Card::Fail, Card::Fail]));
+
+    assert!(is_failure(
+        spec,
+        1,
+        &[Card::Success, Card::Success, Card::Reverse]
+    ));
+    assert!(is_failure(
+        spec,
+        1,
+        &[Card::Fail, Card::Reverse, Card::Reverse]
+    ));
+
+    // TODO: replace with proper 7-player spec
+    let mut with_seven = spec.clone();
+    with_seven.double_fail_mission_four = true;
+
+    assert!(!is_failure(
+        &with_seven,
+        4,
+        &[Card::Success, Card::Success, Card::Success, Card::Fail]
+    ));
+    assert!(!is_failure(
+        &with_seven,
+        4,
+        &[Card::Success, Card::Reverse, Card::Fail, Card::Fail]
+    ));
+    assert!(!is_failure(
+        &with_seven,
+        4,
+        &[Card::Success, Card::Success, Card::Success, Card::Reverse]
+    ));
+    assert!(!is_failure(
+        &with_seven,
+        4,
+        &[Card::Reverse, Card::Reverse, Card::Fail, Card::Success]
+    ));
+
+    assert!(is_failure(
+        &with_seven,
+        4,
+        &[Card::Success, Card::Success, Card::Fail, Card::Fail]
+    ));
+    // This is why reversing mission 4 as Lance is a choice
+    assert!(is_failure(
+        &with_seven,
+        4,
+        &[Card::Success, Card::Success, Card::Reverse, Card::Fail]
+    ));
+    assert!(is_failure(
+        &with_seven,
+        4,
+        &[Card::Reverse, Card::Reverse, Card::Fail, Card::Fail]
+    ));
+}
+
+#[cfg(test)]
+mod test {
+    use futures::executor::block_on;
```

Comment:
Is there a reason to use `futures::executor::block_on` instead of being consistent with `tokio::test`?

Reply:
`block_on` avoids spinning up a Tokio runtime, so it should be faster when we don't need one